QT_Py_Cube/cube.py

```diff
@@ -0,0 +1,185 @@
+            self.done_scrolling = False
+            self.clear = False
+
+        self.__scroll_framebuf(self.word, self.scroll_x_pos, 0)
+        self.__display_top_pixels()
+        self.scroll_x_pos = self.scroll_x_pos + 1
+
+    def clear_cube(self, clear_top=False):
+        if not self.clear:
+            self.pixel_framebuf_sides.fill(0)
+            self.pixel_framebuf_sides.display()
+            self.side_pixels.fill(0)
+            self.side_pixels.show()
+            self.pixel_framebuf_bottom.fill(0)
+            self.pixel_framebuf_bottom.display()
+            if clear_top:
+                self.pixel_framebuf_top.fill(0)
+                self.pixel_framebuf_top.display()
+            self.clear = True
+
+    def upside_down_mode(self):
+        self.clear_cube(True)
+        self.rainbow_sides.animate()
+        now = time.monotonic()
+        self.__bottom_square_animation(now)
+
+    def waiting_mode(self):
+        self.pixel_framebuf_top.pixel(3, 3, self.__convert_to_hex(CYAN))
+        self.pixel_framebuf_top.pixel(4, 4, self.__convert_to_hex(PINK))
+        self.pixel_framebuf_top.display()
+
+    def __bottom_square_animation(self, now):
+        self.pixel_framebuf_bottom.fill(0)
+        color_int = self.__convert_to_hex(CYAN)
+        if now > self.bottom_last + self.bottom_wait:
+            self.__coord_wrap()
+            self.bottom_last = now
+            x, y, w, h = self.bottom_squares[self.bottom_squares_idx]
+            self.pixel_framebuf_bottom.rect(x, y, w, h, color_int)
+            self.pixel_framebuf_bottom.display()
+
+    def __coord_wrap(self):
+        self.bottom_squares_idx = self.bottom_squares_idx + 1
+        if self.bottom_squares_idx >= len(self.bottom_squares):
+            self.bottom_squares_idx = 0
+
+    def __display_top_pixels(self):
+        self.pixel_framebuf_top.fill(0)
+        color_int = self.__convert_to_hex(self.top_pixel_color)
+
+        for coord in self.top_pixel_coords:
+            x, y = coord.split(":")
+            self.pixel_framebuf_top.pixel(int(x), int(y), color_int)
+        self.pixel_framebuf_top.display()
+
+    def __scroll_framebuf(self, word, shift_x, shift_y):
+        self.pixel_framebuf_sides.fill(0)
+
+        color = self.__next_color()
+        color_int = self.__convert_to_hex(color)
+
+        # negate x so that the word can be shown from left to right
+        self.pixel_framebuf_sides.text(word, -shift_x, shift_y, color_int)
+        self.pixel_framebuf_sides.display()
+
+    def __next_color(self):
+        if self.color_idx >= len(self.color_list):
+            self.color_idx = 0
+
+        result = self.color_list[self.color_idx]
+        now = time.monotonic()
+        if now >= self.last_color_time + self.color_wait:
+            self.color_idx = self.color_idx + 1
+            self.last_color_time = now
+
+        return result
+
+    @staticmethod
+    def __convert_to_hex(color):
+        return int('0x%02x%02x%02x' % color, 16)
```

Comment:
A color is an integer, so this `__convert_to_hex()` is not needed: it converts an integer to hex and then back to an integer again. So just use the color name variable (like `CYAN`) and you should not need this routine at all.

Reply:
Have asked the author to review

Reply:
Thanks for the review! I was able to remove a lot of usages of this `__convert_to_hex()` function, but unfortunately still need it in two places -- for some reason, when i pass in just a color (e.g. CYAN) to `PixelFramebuffer.text` or `PixelFramebuffer.rect`, I get the following error:

```
  File "adafruit_framebuf.py", line 318, in rect
  File "adafruit_framebuf.py", line 165, in fill_rect
TypeError: unsupported types for __rshift__: 'tuple', 'int'
```

It's not immediately apparent to me what this error means, but i do seem to get around it by using this convert method. I've sent an updated copy of the cube.py file, happy to discuss if more changes are needed as well, thanks!

Reply:
I was wrong. The colors are tuples: `(R, G, B)` values. You can use those many places, but for for those locations, you can use `int.from_bytes()`, which will be faster than formatting a string:
https://github.com/todbot/circuitpython-tricks#convert-rgb-tuples-to-int-and-back-again
```py
rgb_tuple = (255, 0, 128)
rgb_int = int.from_bytes(rgb_tuple, byteorder='big')
```

Reply:
ooh nice! unfortunately I just tried this:

```
@staticmethod
    def __convert(color):
        return int.from_bytes(color, byteorder='big')
```

and it gave me this error:
```
TypeError: function doesn't take keyword arguments
```

tried with `int.from_bytes(color)`
and got `TypeError: function missing 1 required positional arguments`

and trying it with `int.from_bytes(color, 'big')`
got the error `TypeError: object with buffer protocol required`

I did look at the [python stdtypes documentation](https://docs.python.org/3/library/stdtypes.html) and it does look like what you suggested should have worked, but it also looks like the circuitpython docs say something else:
https://docs.circuitpython.org/en/latest/docs/library/builtins.html?highlight=byteorder#int
> In CircuitPython, byteorder parameter must be positional (this is compatible with CPython).

So I do have to pass _something_ in as a positional parameter, just not sure which object that needs to be?

I also did a search for `int.from_bytes` in this repo, and it looks like there's only two comments referencing this function, to say that the function doesn't support "LE" (and I'm not sure what that means 😅)  

Should I just keep the code as is for now since it seems like it might be the simplest path forward even though it's not the fastest it can be? :D

Screenshot of repo search below:

<img width="759" alt="Screen Shot 2022-09-06 at 10 03 50 PM" src="https://user-images.githubusercontent.com/3520614/188792924-a9ad136d-3052-4842-b7df-ecc168f9f511.png">

Reply:
Thanks so much! Just tested it and it worked -- changes are incoming :)

Thanks for the help, and I'm glad I learned something new this week 🥳